lib/ddtrace/buffer.rb

```diff
@@ -41,31 +41,39 @@ def concat(items)
 
     # Return the current number of stored items.
     def length
-      @mutex.synchronize do
+      synchronize do
         return @items.length
       end
     end
 
     # Return if the buffer is empty.
     def empty?
-      @mutex.synchronize do
+      synchronize do
         return @items.empty?
       end
     end
 
     # Stored items are returned and the local buffer is reset.
     def pop
-      @mutex.synchronize do
+      synchronize do
         drain!
       end
     end
 
     def close
-      @mutex.synchronize do
+      synchronize do
         @closed = true
       end
     end
 
+    def synchronize
```

Comment:
This seems like a re-entrant mutex.
If that's the case, Ruby has a built-in one: https://ruby-doc.org/stdlib-2.5.3/libdoc/monitor/rdoc/Monitor.html (with zero documentation for some reason)

Reply:
Good call, didn't know about this.